File: src/providers/openai-realtime/rules/no-log-raw-message-payloads.ts
import { collectOpenAIRealtimeSocketVarNames, isSocketOnCall } from '../utils.js';

const LOG_METHOD_NAMES = new Set(['info', 'warn', 'error', 'log']);

function isLogCall(node: any): { matches: boolean; calleeProp?: any } {
  if (node?.type !== 'CallExpression') return { matches: false };
  const callee = node.callee;
  if (callee?.type !== 'MemberExpression') return { matches: false };
  const prop = callee.property;
  if (prop?.type !== 'Identifier' || !LOG_METHOD_NAMES.has(prop.name)) return { matches: false };
  return { matches: true, calleeProp: prop };
}

/** True if `argNode` references `paramName` as a whole value (not a narrowed property access). */
function referencesRawParam(argNode: any, paramName: string): boolean {
  if (!argNode) return false;

  // Direct identifier: logger.info(msg)
  if (argNode.type === 'Identifier' && argNode.name === paramName) return true;

  // msg.toString()
  if (
    argNode.type === 'CallExpression' &&
    argNode.callee?.type === 'MemberExpression' &&
    argNode.callee.object?.type === 'Identifier' &&
    argNode.callee.object.name === paramName &&
    argNode.callee.property?.type === 'Identifier' &&
    argNode.callee.property.name === 'toString'
  ) {
    return true;
  }

  // Template literal: `... ${msg} ...` or `... ${msg.toString()} ...`
  if (argNode.type === 'TemplateLiteral') {
    return (argNode.expressions ?? []).some((expr: any) => referencesRawParam(expr, paramName));
  }

  return false;
}

function findCallExpressions(node: any, out: any[], depth = 0): void {
  if (!node || typeof node !== 'object' || depth > 40) return;
  if (Array.isArray(node)) {
    for (const n of node) findCallExpressions(n, out, depth + 1);
    return;
  }
  if (node.type === 'CallExpression') out.push(node);
  for (const key of Object.keys(node)) {
    if (key === 'parent' || key === 'loc' || key === 'range') continue;
    const val = node[key];
    if (val && typeof val === 'object') findCallExpressions(val, out, depth + 1);
  }
}

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'Raw OpenAI Realtime message payloads must not be logged verbatim',
      category: 'security',
      cwe: 'CWE-532',
      docsUrl: 'https://developers.openai.com/api/docs/guides/realtime',
      rationale:
        'Every inbound Realtime WebSocket message can include response.audio.delta payloads (base64-encoded live call audio) and, once transcription is wired up, transcript text. Logging the raw message object or string verbatim writes a durable, unredacted record of live conversation content into whatever log sink the application ships to.',
      recommended: true,
    },
    messages: {
      rawPayloadLogged:
        'A raw OpenAI Realtime message is logged verbatim here, which can include live call audio or transcript content.',
    },
    schema: [],
  },
  create(context: any) {
    let socketVarNames = new Set<string>();

    return {
      Program(node: any) {
        socketVarNames = collectOpenAIRealtimeSocketVarNames(node);
      },

      CallExpression(node: any) {
        if (socketVarNames.size === 0) return;
        if (!isSocketOnCall(node, socketVarNames, 'message')) return;

        const handler = node.arguments?.[1];
        if (handler?.type !== 'ArrowFunctionExpression' && handler?.type !== 'FunctionExpression') return;

        const param = handler.params?.[0];
        if (param?.type !== 'Identifier') return;
        const paramName = param.name;

        const calls: any[] = [];
        findCallExpressions(handler.body, calls);

        for (const call of calls) {
          const { matches } = isLogCall(call);
          if (!matches) continue;
          const hasRawArg = (call.arguments ?? []).some((arg: any) => referencesRawParam(arg, paramName));
          if (hasRawArg) {
            context.report({ node: call, messageId: 'rawPayloadLogged' });
          }
        }
      },
    };
  },
};

export const openaiRealtimeNoLogRawMessagePayloadsRule = rule;

File: src/providers/openai-realtime/rules/python/no_log_raw_message_payloads.py
"""openai-realtime-no-log-raw-message-payloads

Parity with the JS rule: every inbound Realtime message can include
`response.audio.delta` payloads (base64-encoded live call audio) and
transcript content. Logging the raw message verbatim (via `async for
message in ws:` / `for message in ws:`) writes a durable, unredacted record
of live conversation content into whatever log sink the app ships to.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import collect_realtime_socket_var_names, is_tracked_socket_ref  # noqa: E402

RULE_KEY = "openai-realtime-no-log-raw-message-payloads"

LOG_METHOD_NAMES = frozenset({"info", "warning", "warn", "error", "debug", "log", "critical", "exception"})
# `print(...)` is a bare-call builtin — Python's idiomatic equivalent of JS's
# member-call `console.log(...)` for dumping a value verbatim.
BARE_LOG_FUNCTION_NAMES = frozenset({"print"})


def _is_log_call(node: ast.AST) -> bool:
    if not isinstance(node, ast.Call):
        return False
    if isinstance(node.func, ast.Attribute) and node.func.attr in LOG_METHOD_NAMES:
        return True
    return isinstance(node.func, ast.Name) and node.func.id in BARE_LOG_FUNCTION_NAMES


def _references_raw_param(arg: ast.AST | None, param_name: str) -> bool:
    if arg is None:
        return False
    if isinstance(arg, ast.Name) and arg.id == param_name:
        return True
    if (
        isinstance(arg, ast.Call)
        and isinstance(arg.func, ast.Attribute)
        and arg.func.attr in {"decode", "strip"}
        and isinstance(arg.func.value, ast.Name)
        and arg.func.value.id == param_name
    ):
        return True
    if isinstance(arg, ast.JoinedStr):
        for v in arg.values:
            if isinstance(v, ast.FormattedValue) and _references_raw_param(v.value, param_name):
                return True
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    socket_var_names = collect_realtime_socket_var_names(tree)
    if not socket_var_names:
        return []

    out: list[dict] = []
    for node in ast.walk(tree):
        if not isinstance(node, (ast.For, ast.AsyncFor)):
            continue
        if not is_tracked_socket_ref(node.iter, socket_var_names) or not isinstance(node.target, ast.Name):
            continue

        param_name = node.target.id
        for stmt in node.body:
            for call in ast.walk(stmt):
                if not _is_log_call(call):
                    continue
                if any(_references_raw_param(a, param_name) for a in call.args):
                    line, col, _, _ = loc(call)
                    out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
